File: examples/observability/dotnet/SemanticKernelExample.cs
using System.Text.Json;
using Cleargate;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;

/// <summary>
/// Semantic Kernel observability example with ClearGate.
/// Uses SemanticKernelFilter to auto-capture SK function invocations
/// and InstrumentedChatCompletionService to auto-capture LLM calls.
/// Requires: Ollama running locally with qwen3:4b model.
/// </summary>
static class SemanticKernelExample
{
    public static async Task RunAsync()
    {
        using var filter = new SemanticKernelFilter("sk-example", storePath: "sqlite://cleargate_runs.db?mode=rwc");

        var builder = Kernel.CreateBuilder();

        // Ollama exposes an OpenAI-compatible API
#pragma warning disable SKEXP0010
        builder.AddOpenAIChatCompletion(
            modelId: "qwen3:4b",
            apiKey: "ollama",
            endpoint: new Uri("http://localhost:11434/v1")
        );
#pragma warning restore SKEXP0010

        var kernel = builder.Build();

        // Add ClearGate filters for auto-capture of function invocations and prompts
        kernel.FunctionInvocationFilters.Add(filter);
        kernel.PromptRenderFilters.Add(filter);

        // Wrap the chat service to auto-capture LLM request/response data
        var rawChatService = kernel.GetRequiredService<IChatCompletionService>();
        var chatService = filter.Instrument(rawChatService, model: "qwen3:4b", provider: "ollama");

        var history = new ChatHistory();
        history.AddUserMessage("What is the capital of France? Answer briefly.");

        var result = await chatService.GetChatMessageContentsAsync(history, kernel: kernel);
        Console.WriteLine($"Response: {result[0].Content}");

        filter.Finish();

        Console.WriteLine($"\nRun ID: {filter.RunId}");
        Console.WriteLine($"Run data: {filter.GetRunData()}");

        // Print captured events
        var eventsJson = filter.GetEvents();
        if (eventsJson != null)
        {
            var events = JsonSerializer.Deserialize<JsonElement[]>(eventsJson);
            Console.WriteLine($"\nCaptured {events?.Length ?? 0} events");

            if (events != null)
            {
                for (var i = 0; i < events.Length; i++)
                {
                    Console.WriteLine($"\n--- Event {i + 1} ---");
                    Console.WriteLine(JsonSerializer.Serialize(events[i], new JsonSerializerOptions { WriteIndented = true }));
                }
            }
        }

        Console.WriteLine("\nRun data persisted to cleargate_runs.db");
    }
}

File: examples/observability/python/semantic_kernel_example.py
"""Semantic Kernel observability example with ClearGate.

Demonstrates auto-capture of SK function invocations
using CleargateSKFilter.

Requires: Ollama running locally with qwen3:4b model.
"""

import asyncio
from openai import AsyncOpenAI
from semantic_kernel import Kernel
from semantic_kernel.connectors.ai.open_ai import OpenAIChatCompletion, OpenAIChatPromptExecutionSettings
from semantic_kernel.contents import ChatHistory
from cleargate.semantic_kernel import CleargateSKFilter


async def main():
    with CleargateSKFilter("semantic-kernel-example", store_path="sqlite://cleargate_runs.db?mode=rwc") as sk_filter:
        kernel = Kernel()

        # Ollama exposes an OpenAI-compatible API — pass a pre-configured client
        client = AsyncOpenAI(api_key="ollama", base_url="http://localhost:11434/v1")
        chat_service = OpenAIChatCompletion(
            ai_model_id="qwen3:4b",
            async_client=client,
        )
        kernel.add_service(chat_service)

        # Add ClearGate filter for auto-capture
        kernel.add_filter("function_invocation", sk_filter.function_invocation_filter)

        chat = ChatHistory()
        chat.add_user_message("What is the capital of France? Answer briefly.")

        settings = OpenAIChatPromptExecutionSettings(ai_model_id="qwen3:4b")
        result = await sk_filter.invoke_chat(
            chat_service,
            chat_history=chat,
            settings=settings,
            kernel=kernel,
        )
        print(f"Response: {result}")

    events = sk_filter.get_events()
    print(f"\nCaptured {len(events)} events")
    print(f"Run ID: {sk_filter.run_id}")


if __name__ == "__main__":
    asyncio.run(main())
